File: src/api/sse/connection_manager.py
import asyncio
from typing import Dict, Set

import orjson
import structlog

logger = structlog.get_logger("sse.connection_manager")


class SSEConnectionManager:
    """
    O(1) routing engine for Server-Sent Events.
    Maintains memory-bounded queues for clients to prevent OOM on slow consumers.
    """

    __slots__ = (
        "_queues",
        "_topic_subscribers",
        "_client_topics",
        "_queue_size",
    )

    def __init__(self, queue_size: int = 100):
        self._queue_size = queue_size
        self._queues: Dict[str, asyncio.Queue] = {}  # client_id → Queue
        self._topic_subscribers: Dict[str, Set[str]] = {}  # topic → {client_id}
        self._client_topics: Dict[str, Set[str]] = {}  # client_id → {topic}

    async def connect(self, client_id: str) -> asyncio.Queue:
        """Allocate a new strict-bounded queue for the client."""
        from fastapi import HTTPException

        from config.provider import GlobalConfigProvider

        config = GlobalConfigProvider().get_config()
        if self.active_count >= config.sse.max_connections:
            logger.warning("SSE connection rejected", reason="max_connections reached")
            raise HTTPException(status_code=503, detail="Server at maximum capacity")

        q = asyncio.Queue(maxsize=self._queue_size)
        self._queues[client_id] = q
        self._client_topics[client_id] = set()
        logger.debug("SSE client connected", client_id=client_id)
        return q

    def disconnect(self, client_id: str) -> None:
        """Clean up all mappings for the client."""
        self._queues.pop(client_id, None)
        topics = self._client_topics.pop(client_id, set())
        for topic in topics:
            if topic in self._topic_subscribers:
                self._topic_subscribers[topic].discard(client_id)
                if not self._topic_subscribers[topic]:
                    del self._topic_subscribers[topic]
        logger.debug("SSE client disconnected", client_id=client_id)

    def subscribe(self, client_id: str, topic: str) -> None:
        """O(1) mapping of topic to client queue."""
        if client_id not in self._queues:
            return

        self._client_topics[client_id].add(topic)
        if topic not in self._topic_subscribers:
            self._topic_subscribers[topic] = set()
        self._topic_subscribers[topic].add(client_id)

    async def publish(self, topic: str, event_type: str, data: str) -> None:
        """
        Push pre-serialized event to all subscribed client queues.
        If a client is too slow, drops their oldest message.
        Non-blocking, O(1) loop.
        """
        subscribers = self._topic_subscribers.get(topic)
        if not subscribers:
            return

        message = f"event: {event_type}\ndata: {data}\n\n"

        for client_id in subscribers:
            q = self._queues.get(client_id)
            if not q:
                continue

            try:
                q.put_nowait(message)
            except asyncio.QueueFull:
                # Client is downloading slower than events are arriving.
                # Drop their oldest event to make room and protect server memory.
                try:
                    q.get_nowait()
                    q.put_nowait(message)
                except (asyncio.QueueEmpty, asyncio.QueueFull):
                    pass

    async def publish_mutation(
        self,
        module: str,
        resource: str,
        target: str,
        action: str,
        details: dict,
    ) -> None:
        """Helper to serialize a DB/FS mutation once and push to specific + wildcard topics."""
        # 1. Pre-serialize once (Zero-copy payload design)
        payload = orjson.dumps(
            {
                "action": action,
                "module": module,
                "resource": resource,
                "target": target,
                "details": details,
            }
        ).decode()

        # 2. Push to both exact match and wildcard
        specific_topic = f"{module}:{resource}:{target}"
        wildcard_topic = f"{module}:{resource}:*"

        # publish handles the internal loop instantly
        await self.publish(specific_topic, "mutation", payload)
        await self.publish(wildcard_topic, "mutation", payload)

    @property
    def active_count(self) -> int:
        return len(self._queues)

    @property
    def topic_count(self) -> int:
        return len(self._topic_subscribers)


# Singleton
sse_mgr = SSEConnectionManager()

File: src/api/sse/connection_manager.rs
use std::collections::{HashMap, HashSet};
use tokio::sync::mpsc;
use tokio::sync::RwLock;
use axum::response::sse::Event;
use once_cell::sync::Lazy;

use crate::config::loader::ConfigManager;

#[derive(Clone, Default)]
pub struct SSEClientScopes {
    pub db_scope: Vec<String>,
    pub fs_scope: Vec<String>,
    pub full_admin: bool,
}

pub struct SSEConnectionManager {
    connections: RwLock<HashMap<String, mpsc::Sender<Event>>>,
    subscriptions: RwLock<HashMap<String, HashSet<String>>>,
    topic_subscribers: RwLock<HashMap<String, HashSet<String>>>,
}

impl SSEConnectionManager {
    pub fn new() -> Self {
        Self {
            connections: RwLock::new(HashMap::new()),
            subscriptions: RwLock::new(HashMap::new()),
            topic_subscribers: RwLock::new(HashMap::new()),
        }
    }

    pub async fn connect(&self, client_id: &str) -> mpsc::Receiver<Event> {
        let config = ConfigManager::get();
        let queue_size = config.sse.queue_size as usize;

        // Use bounded channel to apply backpressure if client is slow
        let (tx, rx) = mpsc::channel(queue_size);

        self.connections.write().await.insert(client_id.to_string(), tx);
        self.subscriptions.write().await.insert(client_id.to_string(), HashSet::new());
        println!("SSE connected: {}", client_id);

        rx
    }

    pub async fn disconnect(&self, client_id: &str) {
        let subs = self.subscriptions.write().await.remove(client_id).unwrap_or_default();
        let mut topic_subs = self.topic_subscribers.write().await;
        for topic in subs {
            if let Some(clients) = topic_subs.get_mut(&topic) {
                clients.remove(client_id);
            }
        }
        self.connections.write().await.remove(client_id);
        println!("SSE disconnected: {}", client_id);
    }

    pub async fn subscribe(&self, client_id: &str, topic: &str) {
        let mut subs_map = self.subscriptions.write().await;
        let subs = subs_map.entry(client_id.to_string()).or_default();
        subs.insert(topic.to_string());
        drop(subs_map);

        self.topic_subscribers.write().await
            .entry(topic.to_string())
            .or_default()
            .insert(client_id.to_string());
    }

    pub async fn publish(&self, topic: &str, event_type: &str, payload: String) {
        let topic_subs = self.topic_subscribers.read().await;
        let connections = self.connections.read().await;

        if let Some(subscribers) = topic_subs.get(topic) {
            let event = Event::default().event(event_type).data(payload);
            for cid in subscribers {
                if let Some(sender) = connections.get(cid) {
                    // Try to send, if full or closed, we ignore.
                    // Backpressure means we drop events for this slow client if their queue is full.
                    let _ = sender.try_send(event.clone());
                }
            }
        }
    }

    pub async fn active_count(&self) -> usize {
        self.connections.read().await.len()
    }

    pub async fn topic_count(&self) -> usize {
        self.topic_subscribers.read().await.len()
    }
}

pub static SSE_MGR: Lazy<SSEConnectionManager> = Lazy::new(SSEConnectionManager::new);
